Integrate 'uv' for enhanced package management

arash77 opened the pull request:
Implement [uv](https://docs.astral.sh/uv/) to improve installation speed and streamline virtual environment creation across scripts. This change ensures faster package management and eliminates the need for manual pip upgrades.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. Try running Galaxy, tests, updates, or anything related to pip packages and see how quickly it fetches them.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

arash77 committed:
Enhance package management by integrating 'uv' for faster installation and virtual environment creation across scripts

nsoranzo commented on test/release.sh:
```diff
@@ -131,8 +131,21 @@ function make_forks() {
+            export PATH="$HOME/.local/bin:$PATH"
+        fi
+        
+        log_exec uv venv "$VENV"
```
What is the advantage of using `uv venv` in this case?

arash77 replied:
It makes environment creation faster.

arash77 commented:
@nsoranzo In [update.sh](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/dependencies/update.sh#L34C28-L34C42) we are already forcing UV installation. Should we move the installation of UV to one level higher in [the CI job](https://github.com/galaxyproject/galaxy/blob/dev/.github/workflows/dependencies.yaml#L19) and instead use the system UV if it exists in the bash code?

```python
if command -v uv >/dev/null; then
    uv="$(command -v uv)"
else
    uv_venv=$(mktemp -d "${TMPDIR:-/tmp}/uv_venv.XXXXXXXXXX")
    python3 -m venv "${uv_venv}"
    "${uv_venv}/bin/python" -m pip install uv
    uv="${uv_venv}/bin/uv"
fi
```

arash77 committed:
Makes uv optional with pip fallbacks

arash77 committed:
Integrate 'uv' installation across workflow files for improved package management

arash77 committed:
Streamlines wheel + conditional dependency setup

arash77 committed:
Remove minimum required version for 'uv'

arash77 committed:
Apply suggestions from code review

Using `uv tool install` instead of `uv pip install --system`

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

arash77 committed:
Upgrade 'uv' setup action to version 7 across all workflow files

mvdbeek reviewed (approved):
Looks good to me, @nsoranzo do you want to take another look ?

nsoranzo commented on packages/package-build-install.sh:
```diff
@@ -65,15 +65,28 @@ while read package; do
-            "${VENV}/bin/pip" install -r <(grep -v test-requirements.txt dev-requirements.txt)
+            if command -v uv >/dev/null; then
+                uv venv "$VENV"
+                uv pip install --python "${VENV}/bin/python" -r <(grep -v test-requirements.txt dev-requirements.txt)
```
```suggestion
                VIRTUAL_ENV="${VENV}" uv pip install -r <(grep -v test-requirements.txt dev-requirements.txt)
```

nsoranzo commented on packages/test.sh:
```diff
@@ -36,12 +36,23 @@ cd "$(dirname "$0")"
+        uv pip install -r ../lib/galaxy/dependencies/pinned-typecheck-requirements.txt
+    else
+        pip install -r ../lib/galaxy/dependencies/pinned-typecheck-requirements.txt
+    fi
```
Can we define a `PIP` variable to either `uv pip` or `pip` and reuse that throughout the script to avoid duplicating every command?

nsoranzo commented on test/release.sh:
```diff
@@ -131,8 +131,12 @@ function make_forks() {
-        log_exec python3 -m venv "$VENV"
-        log_exec "${VENV}/bin/pip" install wheel packaging
+        if command -v uv >/dev/null; then
+            log_exec uv venv "$VENV"
```
Isn't this branch missing the installation of `wheel` and `packaging`?

arash77 committed:
Update UV on running update_lint_requirements.sh

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

arash77 committed:
Refactor virtual environment setup to use VIRTUAL_ENV for pip installation

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

arash77 committed:
Enhance ensure_prereqs function to use 'uv' for pip list if available

arash77 committed:
Add pip installation for wheel and packaging in create_venv function using 'uv'

arash77 committed:
Refactor test.sh to unify pip installation commands using PIP variable for improved readability and maintainability

nsoranzo commented:
> > The "Test Galaxy packages" errors are different than usual, not sure if it's related to this PR.
> 
> If you rebase on top of the latest dev, the package tests should pass (and it's a problem if they are not).

Except fastapi has just released 0.120.1 which is still broken, so the `get_openapi_path` will fail again.

> > The "Test Galaxy packages" errors are different than usual, not sure if it's related to this PR.
> 
> I think the problem is that when using uv, we need to install without resolving dependencies from the lock file or pulling them from PyPI, since we're installing in dependency order from local packages

We still need to install the deps from PyPI, it's only the Galaxy packages that are installed from local.

nsoranzo committed:
Save whole uv path in PIP env variable

in case it's installed inside a virtual environment.

nsoranzo committed:
Cleanup formatting of pyproject.toml files

nsoranzo committed:
Workaround https://github.com/python/mypy/issues/13392

`uv pip` installs directories as editable packages by default.

mypy does not recognize that libraries installed as editable using
setuptools are fully typed even though the `py.typed` file is
correctly included.

This combination results in spurious mypy errors in package tests if
packages are installed via `uv pip`.

See also:
https://github.com/astral-sh/uv/issues/3898
https://github.com/pypa/setuptools/issues/3518

arash77 commented:
Thanks for pushing the fix! @nsoranzo

nsoranzo committed:
Remove last traces of test-requirements.txt

nsoranzo committed:
Disallow also fastapi 0.120.1 and 0.120.2

nsoranzo committed:
Reduce command duplications

nsoranzo committed:
Install uv into ``$GALAXY_CONDA_ENV``

nsoranzo committed:
Shellcheck fixes

jmchilton commented:
I'm sure I introduced some conflicts and I'm sorry about that. This is a pretty exciting step forward though - thank you so much for the implementation @arash77 and thanks for the detailed review @nsoranzo.
